64 bits OP, impl V_ADDC_U32 & V_MAD_U64_U32

viniciuslrangel opened the pull request:
~Still need to implement SSA rewrite auto cast between 64 and 32 bits~

Fixed op src being wrongly forced to float

viniciuslrangel committed:
impl V_ADDC_U32 & V_MAD_U64_U32

viniciuslrangel committed:
shader recompiler: add 64 bits version to get register / GetSrc

viniciuslrangel committed:
fix V_ADDC_U32 carry

viniciuslrangel committed:
shader recompiler: removed automatic conversion to force_flt in GetSRc

viniciuslrangel committed:
shader recompiler: auto cast between u32 and u64 during ssa pass

viniciuslrangel committed:
shader recompiler: fix SetVectorReg64 & standardize switches-case

raphaelthegreat commented on src/shader_recompiler/frontend/translate/vector_alu.cpp:
```diff
@@ -92,6 +92,26 @@ void Translator::V_ADD_I32(const GcnInst& inst) {
+    const auto src0 = GetSrc<IR::U32>(inst.src[0]);
+    const auto src1 = GetSrc<IR::U32>(inst.src[1]);
+
+    IR::U1 scarry;
```
Is this really a U1? It makes more sense for the shader to use lower part of VCC mask as U32 carry

red-prig replied:
> Is this really a U1? It makes more sense for the shader to use lower part of VCC mask as U32 carry

Setting in VCC, each bit corresponds to the thread number in wavefront [0..63] that is, if you do not yet emulate all wavefront functionality, you can, for example, set the first bit

raphaelthegreat replied:
Yeah a lot of instructions use it for per thread bit storage like compares. Some instructions can also use the low 32-bits as temporary storage though

raphaelthegreat commented on src/shader_recompiler/frontend/translate/vector_alu.cpp:
```diff
@@ -294,6 +314,20 @@ void Translator::V_SUBREV_I32(const GcnInst& inst) {
+    result = ir.IAdd(ir.UConvert(64, result), src2);
+
+    const IR::VectorReg dst_reg{inst.dst[0].code};
+    ir.SetVectorReg64(dst_reg, result);
```
Docs say this also sets VCC but it is unclear to me if it is full 64bit register or only lower 32bits

raphaelthegreat commented on src/shader_recompiler/ir/ir_emitter.cpp:
```diff
@@ -145,34 +145,64 @@ void IREmitter::SetThreadBitScalarReg(IR::ScalarReg reg, const U1& value) {
 
+template <>
+U64 IREmitter::GetScalarReg(IR::ScalarReg reg) {
+    return Inst<U64>(Opcode::GetScalarRegister, reg, Imm32(64));
```
This is more complicated than necessary I think. You can avoid modifying the IR instruction by loading 2 scalar/vector registers and combining them with PackUint2x32. So this instruction is kept clean

raphaelthegreat replied:
Also IREmitter should not handle this at all I think. Have Translator do this in 64-bit version of GetSrc

raphaelthegreat replied:
This also depends on how guest shader uses 64-bit arithmetic. For single instances in some instructions I feel this will be a simpler way to achieve the result in the current system. Otherwise if a larger section of the shader performs U64 arithmetic it might be worth investigating other solutions. Probably a constant propagation pass that detects this Pack -> Unpack -> Pack patterns and preserves the original U64 value

viniciuslrangel committed:
shader translate: fix overflow detection in V_ADD_I32

use vcc lo instead of vcc thread bit

viniciuslrangel committed:
shader recompiler: more 64-bit work

- removed bit_size parameter from Get[Scalar/Vector]Register
- add BitwiseOr64
- add SetDst64 as a replacement for SetScalarReg64 & SetVectorReg64
- add GetSrc64 for 64-bit value

raphaelthegreat commented on src/shader_recompiler/ir/passes/ssa_rewrite_pass.cpp:
```diff
@@ -310,7 +310,8 @@ class Pass {
 };
 
-void VisitInst(Pass& pass, IR::Block* block, IR::Inst& inst) {
+void VisitInst(Pass& pass, IR::Block* block, const IR::Block::iterator& iter) {
```
These refactors are not needed now it seems

viniciuslrangel committed:
shader recompiler: add V_MAD_U64_U32 vcc output

- add V_MAD_U64_U32 vcc output
- ILessThan for 64-bits

raphaelthegreat commented on src/shader_recompiler/ir/opcodes.inc:
```diff
@@ -289,6 +293,9 @@ OPCODE(ConvertF64S32,                                       F64,            U32,
 OPCODE(ConvertF64U32,                                       F64,            U32,                                                                            )
 OPCODE(ConvertF32U16,                                       F32,            U16,                                                                            )
 OPCODE(ConvertU16U32,                                       U16,            U32,                                                                            )
+OPCODE(ConvertU64U32,                                       U64,            U32,                                                                            )
```
Are these opcodes needed anymore?

raphaelthegreat commented on src/shader_recompiler/frontend/translate/vector_alu.cpp:
```diff
@@ -92,6 +93,30 @@ void Translator::V_ADD_I32(const GcnInst& inst) {
+    const IR::VectorReg dst_reg{inst.dst[0].code};
+    ir.SetVectorReg(dst_reg, result);
+
+    IR::U1 less_src0 = ir.ILessThan(result, src0, false);
```
```suggestion
    const IR::U1 less_src0 = ir.ILessThan(result, src0, false);
```
ditto for other small cases of const correctness

raphaelthegreat reviewed (commented):
Nice work only small nits left

viniciuslrangel committed:
shader recompiler: removed unnecessary changes & missing consts

viniciuslrangel committed:
shader_recompiler: Add s64 type in constant propagation
